File: WinForms/CS/TreeListMainDemo/Modules/Clipboard.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.IO;
using System.Linq;
using System.Text;
using System.Windows.Forms;
using DevExpress.Export;
using DevExpress.Spreadsheet;

namespace DevExpress.XtraTreeList.Demos {
    public partial class ClipboardFormats : DevExpress.XtraTreeList.Demos.TutorialControl {
        public override TreeList MainControl {
            get { return treeList1; }
        }
        protected override string[] WhatsThisCodeFileNames { get { return new string[] { "TreeListMainDemo\\Modules\\Clipboard" }; } }
        protected override string WhatsThisXMLFileName { get { return "Clipboard"; } }
        public ClipboardFormats() {
            InitializeComponent();
            Init();
        }
        void Init() {
            propertyGridControl.SelectedObject = treeList1.OptionsClipboard;
            propertyGridControl.ExpandAllRows();
            gridViewPropertiesLayoutGroup.Expanded = false;
        }

        internal void PasteClipboardData() {
            PasteHTMLFormat();
            PasteXLSFormat();
            PasteRTFFormat();
        }
        //<richEditControl>
        void PasteRTFFormat() {
            richEditControl.Document.RtfText = string.Empty;
            richEditControl.Document.Text = string.Empty;
            try {
                if(treeList1.OptionsClipboard.ClipboardMode == Export.ClipboardMode.Formatted) {
                    string rtfText = DevExpress.Data.Utils.SafeClipboardWin.Instance.GetData(DataFormats.Rtf) as string;
                    richEditControl.Document.AppendRtfText(rtfText);
                }
                else {
                    string text = DevExpress.Data.Utils.SafeClipboardWin.Instance.GetData(DataFormats.UnicodeText) as string;
                    richEditControl.Document.AppendText(text);
                }
            }
            catch { }
        }
        //</richEditControl>
        protected override void OnUseEmptyRootPadding() { }
        //<spreadsheetControl1>
        void PasteXLSFormat() {
            spreadsheetControl1.ActiveWorksheet.Clear(spreadsheetControl1.ActiveWorksheet.GetDataRange());
            try {
                if(treeList1.OptionsClipboard.ClipboardMode == ClipboardMode.Formatted) {
                    var stream = DevExpress.Data.Utils.SafeClipboardWin.Instance.GetData("Biff8") as Stream;
                    if(stream != null)
                        spreadsheetControl1.LoadDocument(stream, DocumentFormat.Xls);
                }
                else {
                    int nextRow = 0;
                    string text = DevExpress.Data.Utils.SafeClipboardWin.Instance.GetData(DataFormats.UnicodeText) as string;
                    foreach(var line in text.Split('\n')) {
                        spreadsheetControl1.ActiveWorksheet.Import(line.Split('\t'), nextRow, 0, false);
                        nextRow++;
                    }
                }
            }
            catch { }
            spreadsheetControl1.ActiveWorksheet.DefaultColumnWidthInPixels = spreadsheetControl1.Width / treeList1.Columns.Count;
        }
        //</spreadsheetControl1>
        //<webBrowser1>
        void PasteHTMLFormat() {
            if(treeList1.OptionsClipboard.ClipboardMode == ClipboardMode.Formatted) {
                try {
                    string html = DevExpress.Data.Utils.SafeClipboardWin.Instance.GetData(DataFormats.Html) as string;
                    if(string.IsNullOrEmpty(html)) {
                        webBrowser1.DocumentText = string.Empty;
                        return;
                    }
                    html = html.Remove(0, html.Substring(0, html.IndexOf("<html", StringComparison.OrdinalIgnoreCase)).Length);
                    webBrowser1.DocumentText = html;
                }
                catch { }
            }
            else {
                webBrowser1.DocumentText = "<html><body><p align=\"center\">Copying data in HTML format is not supported in this Clipboard Mode</p></body></html>";
            }
        }
        //</webBrowser1>
        //<treeList1>
        void simpleButtonCopyPaste_Click(object sender, EventArgs e) {
            treeList1.CopyAndPaste();
        }
        //</treeList1>
        void layoutControl1_GroupExpandChanged(object sender, XtraLayout.Utils.LayoutGroupEventArgs e) {
            if(gridViewPropertiesLayoutGroup.Expanded) {
                splitterItem1.Visibility = XtraLayout.Utils.LayoutVisibility.Always;
            }
            else splitterItem1.Visibility = XtraLayout.Utils.LayoutVisibility.Never;
        }

        void ClipboardFormats_Load(object sender, EventArgs e) {
            treeList1.DataSource = SalesDataGenerator.CreateData();
            treeList1.ExpandAll();
            treeList1.ForceInitialize();
            treeList1.BestFitColumns();
            treeList1.SelectCells(treeList1.Nodes[0], colRegion, treeList1.Nodes[0].LastNode, colMarketShare);
        }

        private void treeList_ShowingEditor(object sender, CancelEventArgs e) {
            e.Cancel = true;
        }
    }
    public class ClipboardTreeListControl : TreeList {
        ClipboardFormats userControlCore;
        internal ClipboardFormats ClipboardFormats {
            get {
                if(userControlCore == null)
                    userControlCore = Parent.Parent as ClipboardFormats;
                return userControlCore;
            }
        }
        public void CopyAndPaste() {
            OnKeyDown(new KeyEventArgs(Keys.Control | Keys.C));
        }
        protected override void OnKeyDown(KeyEventArgs e) {
            base.OnKeyDown(e);
            if(e.Control && e.KeyCode == Keys.C) {
                ClipboardFormats.PasteClipboardData();
            }
        }
    }
}

File: WinForms/VB/TreeListMainDemo/Modules/Clipboard.vb
Imports System
Imports System.ComponentModel
Imports System.Drawing
Imports System.IO
Imports System.Windows.Forms
Imports DevExpress.Export
Imports DevExpress.Spreadsheet

Namespace DevExpress.XtraTreeList.Demos

    Public Partial Class ClipboardFormats
        Inherits TutorialControl

        Public Overrides ReadOnly Property MainControl As TreeList
            Get
                Return treeList1
            End Get
        End Property

        Protected Overrides ReadOnly Property WhatsThisCodeFileNames As String()
            Get
                Return New String() {"TreeListMainDemo\Modules\Clipboard"}
            End Get
        End Property

        Protected Overrides ReadOnly Property WhatsThisXMLFileName As String
            Get
                Return "Clipboard"
            End Get
        End Property

        Public Sub New()
            InitializeComponent()
            Init()
        End Sub

        Private Sub Init()
            propertyGridControl.SelectedObject = treeList1.OptionsClipboard
            propertyGridControl.ExpandAllRows()
            gridViewPropertiesLayoutGroup.Expanded = False
        End Sub

        Friend Sub PasteClipboardData()
            PasteHTMLFormat()
            PasteXLSFormat()
            PasteRTFFormat()
        End Sub

        '<richEditControl>
        Private Sub PasteRTFFormat()
            richEditControl.Document.RtfText = String.Empty
            richEditControl.Document.Text = String.Empty
            Try
                If treeList1.OptionsClipboard.ClipboardMode = ClipboardMode.Formatted Then
                    Dim rtfText As String = TryCast(DevExpress.Data.Utils.SafeClipboardWin.Instance.GetData(DataFormats.Rtf), String)
                    richEditControl.Document.AppendRtfText(rtfText)
                Else
                    Dim text As String = TryCast(DevExpress.Data.Utils.SafeClipboardWin.Instance.GetData(DataFormats.UnicodeText), String)
                    richEditControl.Document.AppendText(text)
                End If
            Catch
            End Try
        End Sub

        '</richEditControl>
        Protected Overrides Sub OnUseEmptyRootPadding()
        End Sub

        '<spreadsheetControl1>
        Private Sub PasteXLSFormat()
            spreadsheetControl1.ActiveWorksheet.Clear(spreadsheetControl1.ActiveWorksheet.GetDataRange())
            Try
                If treeList1.OptionsClipboard.ClipboardMode = ClipboardMode.Formatted Then
                    Dim stream = TryCast(DevExpress.Data.Utils.SafeClipboardWin.Instance.GetData("Biff8"), Stream)
                    If stream IsNot Nothing Then spreadsheetControl1.LoadDocument(stream, DocumentFormat.Xls)
                Else
                    Dim nextRow As Integer = 0
                    Dim text As String = TryCast(DevExpress.Data.Utils.SafeClipboardWin.Instance.GetData(DataFormats.UnicodeText), String)
                    For Each line In text.Split(Microsoft.VisualBasic.Strings.ChrW(10))
                        spreadsheetControl1.ActiveWorksheet.Import(line.Split(Microsoft.VisualBasic.Strings.ChrW(9)), nextRow, 0, False)
                        nextRow += 1
                    Next
                End If
            Catch
            End Try

            spreadsheetControl1.ActiveWorksheet.DefaultColumnWidthInPixels = spreadsheetControl1.Width \ treeList1.Columns.Count
        End Sub

        '</spreadsheetControl1>
        '<webBrowser1>
        Private Sub PasteHTMLFormat()
            If treeList1.OptionsClipboard.ClipboardMode = ClipboardMode.Formatted Then
                Try
                    Dim html As String = TryCast(DevExpress.Data.Utils.SafeClipboardWin.Instance.GetData(DataFormats.Html), String)
                    If String.IsNullOrEmpty(html) Then
                        webBrowser1.DocumentText = String.Empty
                        Return
                    End If

                    html = html.Remove(0, html.Substring(0, html.IndexOf("<html", StringComparison.OrdinalIgnoreCase)).Length)
                    webBrowser1.DocumentText = html
                Catch
                End Try
            Else
                webBrowser1.DocumentText = "<html><body><p align=""center"">Copying data in HTML format is not supported in this Clipboard Mode</p></body></html>"
            End If
        End Sub

        '</webBrowser1>
        '<treeList1>
        Private Sub simpleButtonCopyPaste_Click(ByVal sender As Object, ByVal e As EventArgs)
            treeList1.CopyAndPaste()
        End Sub

        '</treeList1>
        Private Sub layoutControl1_GroupExpandChanged(ByVal sender As Object, ByVal e As XtraLayout.Utils.LayoutGroupEventArgs)
            If gridViewPropertiesLayoutGroup.Expanded Then
                splitterItem1.Visibility = XtraLayout.Utils.LayoutVisibility.Always
            Else
                splitterItem1.Visibility = XtraLayout.Utils.LayoutVisibility.Never
            End If
        End Sub

        Private Sub ClipboardFormats_Load(ByVal sender As Object, ByVal e As EventArgs)
            treeList1.DataSource = SalesDataGenerator.CreateData()
            treeList1.ExpandAll()
            treeList1.ForceInitialize()
            treeList1.BestFitColumns()
            treeList1.SelectCells(treeList1.Nodes(0), colRegion, treeList1.Nodes(0).LastNode, colMarketShare)
        End Sub

        Private Sub treeList_ShowingEditor(ByVal sender As Object, ByVal e As CancelEventArgs)
            e.Cancel = True
        End Sub
    End Class

    Public Class ClipboardTreeListControl
        Inherits TreeList

        Private userControlCore As ClipboardFormats

        Friend ReadOnly Property ClipboardFormats As ClipboardFormats
            Get
                If userControlCore Is Nothing Then userControlCore = TryCast(Parent.Parent, ClipboardFormats)
                Return userControlCore
            End Get
        End Property

        Public Sub CopyAndPaste()
            OnKeyDown(New KeyEventArgs(Keys.Control Or Keys.C))
        End Sub

        Protected Overrides Sub OnKeyDown(ByVal e As KeyEventArgs)
            MyBase.OnKeyDown(e)
            If e.Control AndAlso e.KeyCode = Keys.C Then
                ClipboardFormats.PasteClipboardData()
            End If
        End Sub
    End Class
End Namespace
